# 데이터셋 준비

## import library

In [ ]:
# import dependencies
from IPython.display import display, Javascript, Image
from google.colab.output import eval_js
from google.colab.patches import cv2_imshow
from base64 import b64decode, b64encode
import cv2
import numpy as np
import PIL
import io
import html
import time
import os
import matplotlib.pyplot as plt
%matplotlib inline

## darknet 가져오기

### dataset download

In [ ]:
# clone darknet repo (외부 코드 가져오기 - darknet)
!git clone https://github.com/AlexeyAB/darknet

Cloning into 'darknet'...
remote: Enumerating objects: 15502, done.
remote: Total 15502 (delta 0), reused 0 (delta 0), pack-reused 15502
Receiving objects: 100% (15502/15502), 14.14 MiB | 3.15 MiB/s, done.
Resolving deltas: 100% (10406/10406), done.


In [ ]:
!curl -L "https://app.roboflow.com/ds/XftGFVYL8i?key=cHjE8gjeHB" > roboflow.zip; unzip roboflow.zip; rm roboflow.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   895  100   895    0     0    426      0  0:00:02  0:00:02 --:--:--   426
100 81.0M  100 81.0M    0     0  10.9M      0  0:00:07  0:00:07 --:--:-- 18.1M
Archive:  roboflow.zip
 extracting: README.dataset.txt      
 extracting: README.roboflow.txt     
   creating: test/
 extracting: test/-403-_jpg.rf.bf979ce3b91cb271e1ac97e342724974.jpg  
 extracting: test/-403-_jpg.rf.bf979ce3b91cb271e1ac97e342724974.txt  
 extracting: test/-7-3-_jpg.rf.a1b1d36c0d35c48490a60185e0e50117.jpg  
 extracting: test/-7-3-_jpg.rf.a1b1d36c0d35c48490a60185e0e50117.txt  
 extracting: test/-_-_-72-_jpg.rf.096d01cad6685df0ed3a08d72c1c0cdf.jpg  
 extracting: test/-_-_-72-_jpg.rf.096d01cad6685df0ed3a08d72c1c0cdf.txt  
 extracting: test/-_-_-_-_jpg.rf.9c304688579051bd16eda3ec1d1d9ed2.jpg  
 extracting: test/-_-_-_-_jpg.rf.9c304688579051bd16eda3ec1d1d9ed2.tx

## 2) .data파일 생성

In [ ]:
# mask_data.data 파일 생성
t = open('/content/darknet/data/disposable.data', 'w')

In [ ]:
# roboflow에서 다운로드한 데이터셋의 정보를 살펴보자.
%cat /content/test/_darknet.labels

cup_glass
cup_mug
cup_paper
cup_plastic
straw

In [ ]:
# mask_data.data에 기록
t.write("classes = 5\n")

12

### train, test, valid

* train.txt,valid.txt, test.txt 파일을 생성하여보자.
* 각 텍스트 파일은 train, valid, test 과정에 사용될 이미지 리스트를 포함하고 있다.

In [ ]:
# label 파일 삭제
!rm test/_darknet.labels
!rm train/_darknet.labels
!rm valid/_darknet.labels

In [ ]:
# 이후 이미지 데이터셋을 저장할 경로 지정
path = "/content/darknet/data/images/"
# test.txt 생성
# w 모드로 text 파일을 생성
f = open('/content/darknet/data/test.txt', 'w')
for img in os.listdir('/content/test'):
    if img[-3:]=='jpg':
        # 문자 입력
        f.write(path+img+"\n")
f.close()

# train.txt 생성
# w 모드로 text 파일을 생성
f = open('/content/darknet/data/train.txt', 'w')
for img in os.listdir('/content/train'):
    if img[-3:]=='jpg':
        # 문자 입력
        f.write(path+img+"\n")
f.close()

# valid.txt 생성
# w 모드로 text 파일을 생성
f = open('/content/darknet/data/valid.txt', 'w')
for img in os.listdir('/content/valid'):
    if img[-3:]=='jpg':
        # 문자 입력
        f.write(path+img+"\n")
f.close()

In [ ]:
# mask_data.data에 기록

t.write("train = data/train.txt\n")
t.write("valid = data/valid.txt\n")
t.write("test = data/test.txt\n")

21

* train, test, valid 폴더에 있는 images, labels 데이터를 모두 한 폴더에 모아보자.

In [ ]:
# 폴더 병합
import shutil
  
# parent_folder에 folder_name이 기존에 없을 경우 새로 생성하는 함수 생성
def make_new_folder(folder_name, parent_folder):
      
    # Path
    path = os.path.join(parent_folder, folder_name)
      
    # Parent folder에 새로운 폴더 생성
    try: 
        mode = 0o777

        # Path에 폴더 생성
        os.mkdir(path, mode) 
    except OSError as error: 
        print(error)

In [ ]:
# 현재 디렉토리 지정
%cd /content
current_folder = os.getcwd() 

/content


In [ ]:
current_folder

'/content'

`os.listdir()`
* 디렉토리에 있는 파일명을 리스트로 반환한다.

In [ ]:
# 합칠 폴더 목록
folders = ['test', 'train', 'valid']


# folders에 포함된 폴더 내부 콘텐츠 리스트를 딕셔너리로 저장
content_list = {}
for index, folder in enumerate(folders):
    path = os.path.join(current_folder, folder) # path = content/test -> content/train -> content/valid 
    content_list[folders[index]] = os.listdir(path) # os.listdir = [파일경로1, 파일경로2, ...]

# content_list = filter(lambda a: a != '.ipynb_checkpoints', content_list)
content_list # {'test' : [test filenames], 'train' : [train filenames], 'valid' : [valid filenames]}

{'test': ['cups-50-_jpg.rf.07c1460cb434b71c02eda926a39e3155.jpg',
  'Image_2241_jpg.rf.efb1c45ef9ff9a04d25fe787aa399bc6.jpg',
  '-_-_jpg.rf.46d3a88a541c1031e7d3cc1b7323235f.jpg',
  'man-drinking-bubble-milk-tea-in-pedestrian-street_y_jpg.rf.ca697fc6daa9597463a2eed1311c3691.jpg',
  'three-friends-enjoying-hot-beverage-drinks-together_jpg.rf.dd2863250a64e4776cd881bbf9232d39.txt',
  '-__-odod-_yyth_jpg.rf.15a13f545cd1a5159fa71cb3aff76c1c.jpg',
  '001415_jpg.rf.5cf16717deea766a0df1a8085a8d29cf.jpg',
  '-_jpg.rf.a675d71e609d221429597324e2837688.jpg',
  '4-_jpg.rf.682e7a53a3532eacede5119a50583fa8.jpg',
  '-_-_-_CANVAS__jpg.rf.dd7cb9536d686f32a1330dd72b5f1d97.txt',
  '-_-_-_CANVAS__jpg.rf.dd7cb9536d686f32a1330dd72b5f1d97.jpg',
  'row-of-three-plastic-cups-of-various-fruit-s_jpg.rf.84861bceaad0f38fb6cdd598b447eb9d.txt',
  '1498823_jpg.rf.bbc4528966ebe589f6d3a3679fedb346.jpg',
  'portrait-of-young-businesswoman-with-laptop-computer_yythk_jpg.rf.b4a9b5ffcfd7196446b945b583fa8a00.jpg',
  'cups-19-

In [ ]:
# 새롭게 생성할 폴더명 지정
merge_folder = "images"

# 새롭게 생성할 폴더 디렉토리 지정
merge_folder_path = os.path.join('/content/darknet/data', merge_folder) 

# 새로운 폴더 생성
make_new_folder(merge_folder, '/content/darknet/data')

* images 폴더가 생성되었다.

In [ ]:
merge_folder_path

'/content/darknet/data/images'

In [ ]:
# 모든 파일을 수회하며
for sub_dir in content_list: # {'test' : [test filenames], 'train' : [train filenames], 'valid' : [valid filenames]}
    for contents in content_list[sub_dir]: # [test filenames] -> [train filenames] -> [valid filenames]

        # 파일을 이동시킬 폴더 주소 지정 
        path_to_content = sub_dir + "/" + contents # test/testfilename1, ...

        # 현재 주소 지정
        dir_to_move = os.path.join(current_folder, path_to_content) # content/test/testfilename1, /

        # 파일 이동
        shutil.move(dir_to_move, merge_folder_path)

In [ ]:
# 폴더 컨텐츠 개수 출력
%cd '/content/darknet/'
len(os.listdir('data/images'))

/content/darknet


3734

### 라벨별 개수 확인

In [ ]:
import collections

txt_list = []
label_dict={}

for txt in os.listdir('/content/darknet/data/images'):
    if txt[-3:] == 'txt':
        txt_list.append(txt)
    
label_count = []

for txt in txt_list:
    f = open(f"/content/darknet/data/images/{txt}", 'r')
    lines = f.readlines()
    f.close()  
    for line in lines:
        label_count.append(line[:2])

label_dict=collections.Counter(label_count)
print(label_dict)

Counter({'2 ': 1016, '1 ': 777, '4 ': 775, '0 ': 748, '3 ': 702})


### names

In [ ]:
# ClassNames.names 생성
# w 모드로 name 파일을 생성
f = open('/content/darknet/data/ClassNames.names', 'w')

# 문자 입력
f.write("cup_glass"+"\n")
f.write("cup_mug"+"\n")
f.write("cup_paper"+"\n")
f.write("cup_plastic"+"\n")
f.write("straw")

# 파일 닫기
f.close()

In [ ]:
# mask_data.data에 기록

t.write("names = data/ClassNames.names\n")

30

### backup

* training 중간 결과를 저장할 backup 주소를 생성하여보자.

In [ ]:
# training 중간 결과를 저장할 backup 주소를 생성, disposable.data에 기록
t.write("backup = backup/")
t.close()

## 3) .cfg 파일 수정

* cfg 파일은 neural network의 일반 정보를 저장하는 구성 파일이다.
* 우리가 사용할 YOLOv4의 정보는 yolov4-custom.cfg에 저장되어 있다.
* 마스크 착용 여부를 구분하는 task에 맞추어 yolov4-custom.cfg를 수정하자.

다음과 같이 yolov4-custom.cfg 파일을 수정한다.
* batch = 64
* subdivision = 32
* width = 608
* height = 608
* max_batches = 4000
* steps = 3200, 3600
* yolo layer 마다 classes=5
* yolo layer 앞에 있는 convolutional layer 마다 filers = 30

In [ ]:
# 변경사항이 반영이 되었는지 확인하여보자.
%cat /content/darknet/cfg/yolov4-custom.cfg

[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=32
width=608
height=608
channels=3
momentum=0.949
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 5000
policy=steps
steps=4000,4500
scales=.1,.1

#cutmix=1
mosaic=1

#:104x104 54:52x52 85:26x26 104:13x13 for 416

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=mish

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=mish

[route]
layers = -2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=mish

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activatio

## 4) weights 다운로드


`!wget [옵션] [url]`
* 웹 상의 파일을 다운로드하는 리눅스 명령어이다.
* wget => web get



In [ ]:
cd '/content/darknet/'

/content/darknet


In [ ]:
pwd

'/content/darknet'

In [ ]:
# yolov4 pre-trained weights 다운로드
!wget https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.conv.137

--2022-12-14 15:39:56--  https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.conv.137
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/48bfe500-889d-11ea-819e-c4d182fcf0db?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221214%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221214T153956Z&X-Amz-Expires=300&X-Amz-Signature=6986edd793a684da97e4c7dfc5b44d9e6a4b94d861e68fb51ab7d8d457b4c884&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=75388965&response-content-disposition=attachment%3B%20filename%3Dyolov4.conv.137&response-content-type=application%2Foctet-stream [following]
--2022-12-14 15:39:56--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/48bfe500-889d-11ea-819e-c4d

## 5) darknet 구축

In [ ]:
# GPU, OPENCV, LIBSO을 true로 설정
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
!sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!sed -i 's/LIBSO=0/LIBSO=1/' Makefile

In [ ]:
# 위의 변경된 옵션으로 makefile을 컴파일하면 darknet.py 사용이 가능해진다.
!make

mkdir -p ./obj/
mkdir -p backup
chmod +x *.sh
g++ -std=c++11 -std=c++11 -Iinclude/ -I3rdparty/stb/include -DOPENCV `pkg-config --cflags opencv4 2> /dev/null || pkg-config --cflags opencv` -DGPU -I/usr/local/cuda/include/ -DCUDNN -DCUDNN_HALF -Wall -Wfatal-errors -Wno-unused-result -Wno-unknown-pragmas -fPIC -Ofast -DOPENCV -DGPU -DCUDNN -I/usr/local/cudnn/include -DCUDNN_HALF -fPIC -c ./src/image_opencv.cpp -o obj/image_opencv.o
./src/image_opencv.cpp: In function ‘void draw_detections_cv_v3(void**, detection*, int, float, char**, image**, int, int)’:
./src/image_opencv.cpp:946:23: warning: variable ‘rgb’ set but not used [-Wunused-but-set-variable]
                 float rgb[3];
                       ^~~
./src/image_opencv.cpp: In function ‘void draw_train_loss(char*, void**, int, float, float, int, int, float, int, char*, float, int, int, double)’:
./src/image_opencv.cpp:1147:13: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
             if (iteration_old ==

## 6) custom dataset으로 transfer learning

#### best map@0.50 = %

In [ ]:
# custom dataset으로 YOLO 학습 (아래 코드는 model.fit(x, y) 코드라고 이해하면 된다.)
# data 필요, model 구조와 세부설정값, model weights
!./darknet detector train data/disposable.data cfg/yolov4-custom.cfg yolov4.conv.137 -dont_show -map

스트리밍 출력 내용이 길어서 마지막 5000줄이 삭제되었습니다.
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.811770), count: 4, class_loss = 1.797408, iou_loss = 34.678272, total_loss = 36.475681 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.849286), count: 12, class_loss = 0.916215, iou_loss = 24.057344, total_loss = 24.973558 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.838149), count: 2, class_loss = 0.030158, iou_loss = 2.057184, total_loss = 2.087342 
 total_bbox = 1958320, rewritten_bbox = 0.064137 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.000000), count: 1, class_loss = 0.063830, iou_loss = 0.000000, total_loss = 0.063830 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.894756), count: 2, class_loss = 0.455153, iou_loss = 2.093499, total_loss = 2.548652 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Re

## 7) 학습결과 저장

다음 파일을 다운로드한다.
* ClassNames.names
* test.txt
* train.txt
* valid.txt
* mask_data.data
* yolov4_custom_best.weights

google drive에 직접 저장, gdown 링크 생성 

In [ ]:
# /content/drive/MyDrive/plasticr2에 폴더 생성 후 
shutil.copy('/content/darknet/backup/yolov4-custom_best.weights', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/backup/yolov4-custom_last.weights', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/chart.png', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/data/test.txt', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/data/train.txt', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/data/valid.txt', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/data/disposable.data', '/content/drive/MyDrive/final_608/')
shutil.copy('/content/darknet/data/ClassNames.names', '/content/drive/MyDrive/final_608/')

'/content/drive/MyDrive/label4_cctv_test_re/ClassNames.names'

In [ ]:
cd /content/drive/MyDrive

/content/drive/MyDrive


In [ ]:
shutil.make_archive('disposable', 'zip', '/content/drive/MyDrive/label4_cctv_test_re')

'/content/drive/MyDrive/disposable.zip'

# 테스트

## import library

In [ ]:
import tensorflow as tf

In [ ]:
tf.__version__

'2.9.2'

In [ ]:
# import dependencies
from IPython.display import display, Javascript, Image
from google.colab.output import eval_js
from google.colab.patches import cv2_imshow
from base64 import b64decode, b64encode
import cv2
import numpy as np
import PIL
import io
import html
import time
import os
import matplotlib.pyplot as plt
from google.colab import files

%matplotlib inline

In [ ]:
# clone darknet repo
!git clone https://github.com/AlexeyAB/darknet

fatal: destination path 'darknet' already exists and is not an empty directory.


## 1) train에 사용된 파일 업로드

In [ ]:
cd /content/darknet/data

/content/darknet/data


In [ ]:
# gdown 권한 
!pip install -U --no-cache-dir gdown --pre

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
# 구글 드라이브 파일에서 disposable.data을 colab으로 다운로드 
!gdown 1-3c67XQoIajZxJbgtrjtvsCR3ONkZh7a

Downloading...
From: https://drive.google.com/uc?id=1-3c67XQoIajZxJbgtrjtvsCR3ONkZh7a
To: /content/darknet/data/disposable.zip
100% 237M/237M [00:02<00:00, 115MB/s]


In [ ]:
!unzip -q disposable.zip
# dataset내 dataset.zip파일 삭제...

replace disposable.data? [y]es, [n]o, [A]ll, [N]one, [r]ename: ㅛ
error:  invalid response [ㅛ]
replace disposable.data? [y]es, [n]o, [A]ll, [N]one, [r]ename: ㅛ
error:  invalid response [ㅛ]
replace disposable.data? [y]es, [n]o, [A]ll, [N]one, [r]ename: y


## 2) .cfg 파일 수정

다음과 같이 yolo4-custom.cfg 파일을 수정한다.
* testing 활성화(batch=1, subdivisions=1)
* training 주석처리
* width = 416
* height = 416
* max_batches = 3000
* steps = 2400, 2700
* classes = 2
* filers = 21

In [ ]:
%cat /content/darknet/cfg/yolov4-custom.cfg

[net]
# Testing
batch=1
subdivisions=1
# Training
#batch=64
#subdivisions=16
width=416
height=416
channels=3
momentum=0.949
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 3000
policy=steps
steps=2400,2700
scales=.1,.1

#cutmix=1
mosaic=1

#:104x104 54:52x52 85:26x26 104:13x13 for 416

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=mish

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=mish

[route]
layers = -2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=mish

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=mish

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activatio

In [ ]:
# 파일 확인 
%cat /content/darknet/data/disposable.data

classes = 2
train = data/train.txt
valid = data/valid.txt
test = data/test.txt
names = data/ClassNames.names
backup = backup/

## 3) darknet 구축 (필요시)

In [ ]:
pwd

'/content/darknet/data'

In [ ]:
cd /content/darknet

/content/darknet


In [ ]:
# GPU, OPENCV, LIBSO을 true로 설정
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
!sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!sed -i 's/LIBSO=0/LIBSO=1/' Makefile

In [ ]:
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
!sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!apt update
!apt-get install libopencv-dev

Get:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Hit:2 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease
Ign:3 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Get:4 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease [1,581 B]
Hit:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:6 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:7 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [1,082 kB]
Hit:9 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Get:10 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:12 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease
Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease
Get:1

In [ ]:
# 위의 변경된 옵션으로 makefile을 컴파일하면 darknet.py 사용이 가능해진다.
!make

chmod +x *.sh
g++ -std=c++11 -std=c++11 -Iinclude/ -I3rdparty/stb/include -DOPENCV `pkg-config --cflags opencv4 2> /dev/null || pkg-config --cflags opencv` -DGPU -I/usr/local/cuda/include/ -DCUDNN -DCUDNN_HALF -Wall -Wfatal-errors -Wno-unused-result -Wno-unknown-pragmas -fPIC -Ofast -DOPENCV -DGPU -DCUDNN -I/usr/local/cudnn/include -DCUDNN_HALF -fPIC -c ./src/image_opencv.cpp -o obj/image_opencv.o
./src/image_opencv.cpp: In function ‘void draw_detections_cv_v3(void**, detection*, int, float, char**, image**, int, int)’:
./src/image_opencv.cpp:946:23: warning: variable ‘rgb’ set but not used [-Wunused-but-set-variable]
                 float rgb[3];
                       ^~~
./src/image_opencv.cpp: In function ‘void draw_train_loss(char*, void**, int, float, float, int, int, float, int, char*, float, int, int, double)’:
./src/image_opencv.cpp:1147:13: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
             if (iteration_old == 0)
             ^~
./src/image_

In [ ]:
!./darknet detector demo data/disposable.data cfg/yolov4-custom.cfg data/yolov4-custom_best.weights -dont_show test2.mp4 -out_filename video.avi

 CUDA-version: 11020 (11020), cuDNN: 8.1.1, CUDNN_HALF=1, GPU count: 1  
 CUDNN_HALF=1 
 OpenCV version: 3.2.0
Demo
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 1, batch = 1, time_steps = 1, train = 0 
   layer   filters  size/strd(dil)      input                output
   0 Create CUDA-stream - 0 
 Create cudnn-handle 0 
conv     32       3 x 3/ 1    416 x 416 x   3 ->  416 x 416 x  32 0.299 BF
   1 conv     64       3 x 3/ 2    416 x 416 x  32 ->  208 x 208 x  64 1.595 BF
   2 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   3 route  1 		                           ->  208 x 208 x  64 
   4 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   5 conv     32       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  32 0.177 BF
   6 conv     64       3 x 3/ 1    208 x 208 x  32 ->  208 x 208 x  64 1.595 BF
   7 Shortcut Layer: 4,  wt = 0, wn = 0, outputs: 208 x 208 x  64 0.003 BF
   8 co

In [ ]:
!./darknet detector demo data/disposable.data cfg/yolov4-custom.cfg data/yolov4-custom_best.weights -dont_show test2.mp4 -thresh 0.6 -ext_output -out >results/result.json

 CUDA-version: 11020 (11020), cuDNN: 8.1.1, CUDNN_HALF=1, GPU count: 1  
 OpenCV version: 3.2.0
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
   layer   filters  size/strd(dil)      input                output
   0 conv     32       3 x 3/ 1    416 x 416 x   3 ->  416 x 416 x  32 0.299 BF
   1 conv     64       3 x 3/ 2    416 x 416 x  32 ->  208 x 208 x  64 1.595 BF
   2 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   3 route  1 		                           ->  208 x 208 x  64 
   4 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   5 conv     32       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  32 0.177 BF
   6 conv     64       3 x 3/ 1    208 x 208 x  32 ->  208 x 208 x  64 1.595 BF
   7 Shortcut Layer: 4,  wt = 0, wn = 0, outputs: 208 x 208 x  64 0.003 BF
   8 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   9 route  8 2 	                           ->  208 x 208 x 128 
  10 conv 

In [ ]:
import json
import csv
import pandas as pd

with open('/content/darknet/results/result.json', encoding='utf-8') as inputfile:
    df = pd.read_json(inputfile)

df.to_csv('csvfile.csv', encoding='utf-8', index=False)

ValueError: ignored

In [ ]:
cd /content/darknet

/content/darknet


In [ ]:
!./darknet detector demo data/disposable.data cfg/yolov4-custom.cfg data/yolov4-custom_best.weights -dont_show /content/test4.mp4 -out_filename /content/test4.avi

 CUDA-version: 11020 (11020), cuDNN: 8.1.1, CUDNN_HALF=1, GPU count: 1  
 CUDNN_HALF=1 
 OpenCV version: 3.2.0
Demo
 0 : compute_capability = 750, cudnn_half = 1, GPU: Tesla T4 
net.optimized_memory = 0 
mini_batch = 1, batch = 1, time_steps = 1, train = 0 
   layer   filters  size/strd(dil)      input                output
   0 Create CUDA-stream - 0 
 Create cudnn-handle 0 
conv     32       3 x 3/ 1    416 x 416 x   3 ->  416 x 416 x  32 0.299 BF
   1 conv     64       3 x 3/ 2    416 x 416 x  32 ->  208 x 208 x  64 1.595 BF
   2 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   3 route  1 		                           ->  208 x 208 x  64 
   4 conv     64       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  64 0.354 BF
   5 conv     32       1 x 1/ 1    208 x 208 x  64 ->  208 x 208 x  32 0.177 BF
   6 conv     64       3 x 3/ 1    208 x 208 x  32 ->  208 x 208 x  64 1.595 BF
   7 Shortcut Layer: 4,  wt = 0, wn = 0, outputs: 208 x 208 x  64 0.003 BF
   8 co

In [ ]:
pwd

'/content/darknet'

In [ ]:
cd darknet

/content/darknet


In [ ]:
from IPython.display import HTML
from base64 import b64encode
mp4 = open('test2.mp4','rb').read()
data_url = "data:video/mp4;base64," + b64encode(mp4).decode()
HTML("""
<video width=400 controls>
      <source src="%s" type="video/mp4">
</video>
""" % data_url)

FileNotFoundError: ignored

In [ ]:
cd ..

/content


In [ ]:
avi = open('test4.avi','rb').read()
data_url = "data:video/avi;base64," + b64encode(mp4).decode()
HTML("""
<video width=400 controls>
      <source src="%s" type="video/avi">
</video>
""" % data_url)

## 4) Darknet for Python

파이썬 코드로 YOLO ver 4을 사용하기 위해서는 darkent에서 찾을 수 있는 사전에 구축된 function들을 사용할 것이다. function들을 import하여보자. 
 

In [ ]:
from darknet import *

## 5) Yolov4모델 로드
 

In [ ]:
pwd

'/content/darknet'

In [ ]:
cd /content/darknet

/content/darknet


darknet 폴더 내에 있어야 함

In [ ]:
# YOLOv4 모델을 로드하여보자.
# 커스텀 모델을 사용할 경우 load_network의 1번째, 3번째 인자를 수정해야 한다.
network, class_names, class_colors = load_network("cfg/yolov4-custom.cfg", "data/disposable.data", "backup/yolov4-custom_best.weights")

* darknet.py 파일을 확인해보면 load_network 함수의 1번째 인자는 config file, 2번째 인자는 data file, 3번째 인자는 weights에 해당하는 것을 알 수 있다.
* 또한 load_network 함수는 학습된 모델(=network), class 이름(=class names), class 별 고유 색상(=class colors)를 반환하는 것을 알 수 있다.
* class colors는 이후 boundary box를 이미지 위에 시각화할 때 사용할 색상이다.

In [ ]:
# class의 종류 예시
class_names

['disposable', 'reusable']

In [ ]:
# class color의 예시
class_colors['disposable']

(152, 120, 147)

* 3가지의 값은 각각 B(Black), G(Green), R(Red)에 해당하는 것은 알 수 있다.
* 또한 색상은 랜덤하게 배정이 되었으며, 딕셔너리 형태로 클래스 별 생상 데이터가 저장되었음을 알 수 있다.
* 참고: OpenCV는 일반적인 RGB 순서가 아닌 BGR을 사용한다.

In [ ]:
# 네트워크의 width와 height을 계산
width = network_width(network)
height = network_height(network)

print(width, height)

416 416


`cv2.cvtColor(BGR이미지, cv2.COLOR_BGR2RGB)`

* 컬러 사진을 OpenCV에서는 BGR 순서로 인식하지만 일반적으로 matplotlib 등에서 사진을 출력하기 위하여는 이미지가 RGB 순서로 저장되어 있어야 한다.
* 따라서, 위와 같은 이유등을 위하여 이미지 순서를 RGB로 변환할 때 cvtColor을 사용한다.

`cv2.resize(이미지, dsize, fx, fy, interplolation)`
* 이미지의 크기를 조정한다.
* dsize는 결과 영상의 크기를 나타내며 (width, height)의 튜플 형태를 띈다.
* dsize가 (0,0)일 경우 fx와 fy를 통해 크기를 조정한다. fx와 fy는 스케일 비율을 나타낸다.
* interpolation은 보간법을 지정한다. 보간법은 이미지 확대 또는 축소 과정에서 이미지 픽셀의 유실이 일어났을 때, 주변 픽셀값을 이용하여 이를 채워넣는 방법을 의미한다. cv2.INTER_NEAREST,cv2.INTER_LINEAR, cv2.INTER_CUBIC 등이 있으며, INTER_LINEAR의 방법이 속도가 빠를 뿐더러 성능이 좋기 때문에 가장 많이 사용된다.

In [ ]:
# object detection을 수행하는 darknet_helper 함수를 살펴보자.

def darknet_helper(img, width, height):
  # 이미지를 darknet style로 전처리 -> darknet_image 변수에 저장
  darknet_image = make_image(width, height, 3)
  # input image를 BRG에서 RGB 순서로 변환
  img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
  # input image의 size를 512x512로 변환 (network에서 받아들이는 이미지의 크기가 512x512였기 때문에)
  img_resized = cv2.resize(img_rgb, (width, height),
                              interpolation=cv2.INTER_LINEAR)

  # bounding box의 크기를 조정하기 위하여 이미지 크기 비율을 계산 
  img_height, img_width, _ = img.shape
  width_ratio = img_width/width # 원본 이미지 width/ network width
  height_ratio = img_height/height # 원본 이미지 height/ network height

  # darknet stlye로 전처리된 이미지로 모델을 작동시켜 detection을 수행
  copy_image_from_bytes(darknet_image, img_resized.tobytes())
  # detect_image 함수 실행
  detections = detect_image(network, class_names, darknet_image)
  free_image(darknet_image)

  return detections, width_ratio, height_ratio

* darknet_helper은 이미지 전처리(RGB로 변환, 크기 조정)부터 bounding box의 비율 계산, detection까지 모두 수행하는 함수이다. 